### PR TITLE
Remove `MBEDTLS_ECDH_LEGACY_CONTEXT` config option

### LINKED ISSUE #4404: Remove the `MBEDTLS_ECDH_LEGACY_CONTEXT` config.h option
This option was introduced purely for backwards compatibility. Since the whole point of 3.0 is we can break compatibility, let's remove this option

Work to do: see the next comment!

(Original list for reference, which has been split in multiple issues in the meantime.)

1. Change how restartable ECDH works so that `MBEDTLS_ECP_RESTARTABLE` no longer depends on  `MBEDTLS_ECDH_LEGACY_CONTEXT`. *This step might be tricky or require familiarity with that part of the code.* At the end, one should be able to build with `MBEDTLS_ECP_RESTARTABLE` defined and `MBEDTLS_ECDH_LEGACY_CONTEXT` unset and all tests passing.
2. Adapt the benchmark program so that it no longer depends on that option. Remove the test that's specific to that option in `test_suite_ecdh`.
3. Remove `MBEDTLS_ECDH_LEGACY_CONTEXT` and let the code behave as if it was always disabled.
4. In the end `MBEDTLS_ECDH_LEGACY_CONTEXT` should find no match except in the ChangeLog and the CI should pass.

### PR DESCRIPTION
**Pull request by tom-daubney-arm**
## Description
Fixes #4404.

## Status
**READY**

## Requires Backporting
NO

### CODE REVIEW

File: include/mbedtls/ecdh.h [hunk 1]
```diff
@@ -40,6 +40,39 @@
...
+ * no difference, since all library functions can work with either format,
+ * except that the new format is incompatible with MBEDTLS_ECP_RESTARTABLE.
+
+ * The new format used when this option is disabled is smaller
+ * (56 bytes on a 32-bit platform). In future versions of the library, it
+ * will support alternative implementations of ECDH operations.
+ * The new format is incompatible with applications that access
+ * context fields directly and with restartable ECP operations.
+ *
+ * Define this macro if you enable MBEDTLS_ECP_RESTARTABLE or if you
```
Comment (mpg) [inline include/mbedtls/ecdh.h:59]: I think this paragraph should be removed, as this is no longer a user-configurable option.

File: include/mbedtls/ecdh.h [hunk 2]
```diff
@@ -40,6 +40,39 @@
...
+ * want to access ECDH context fields directly. Otherwise you should
+ * comment out this macro definition.
+ *
+ * This option has no effect if #MBEDTLS_ECDH_C is not enabled.
+ *
+ * \note This configuration option is experimental. Future versions of the
```
Comment (mpg) [inline include/mbedtls/ecdh.h:65]: This paragraph is probably no longer relevant either.

File: include/mbedtls/ecdh.h [hunk 3]
```diff
@@ -40,6 +40,39 @@
 
 #include "mbedtls/ecp.h"
 
+/**
```
Comment (mpg) [inline include/mbedtls/ecdh.h:43]: I don't think that should be a doxygen comment. This is now internal, so there's not need to this to be included in the public API documentation. I suggest removing the second `*` here, and the `\` directives below (or `check-doxy-blocks.pl` will complain).

Also, we should add an explicit warning that this in internal, something to the effect of "don't look at this unless you're a compiler" - ideally we would hide this from the user, but unfortunately if needs to be in the header file as it affects the definition of the context structure.

**Review by mpg — COMMENTED** [review]
This looks pretty good to me! IMO we just need to tune the documentation a bit in order to better reflect the new status of this macro.

File: include/mbedtls/ecdh.h (hunk 1 above)
Comment (tom-daubney-arm) [reply include/mbedtls/ecdh.h:59]: Thanks @mpg ! I wasn't quite sure how to handle the documentation of this bit so I appreciate your comments. I will fix them up and push again.

Comment (tom-daubney-arm) [reply include/mbedtls/ecdh.h]: @mpg - I think I have done what you wanted me to do, please re-review when you get the chance.

**Review by mpg — APPROVED** [review]
Thanks for making the changes I suggested. Looks good to me now.

File: include/mbedtls/config.h [hunk 4]
```diff
@@ -784,34 +784,6 @@
  */
 //#define MBEDTLS_ECP_RESTARTABLE
 
-/**
```
Comment (ronald-cron-arm) [inline include/mbedtls/config.h:787]: In the documentation of MBEDTLS_ECP_RESTARTABLE there is a mention to MBEDTLS_ECDH_LEGACY_CONTEXT which I think was wrong and should be removed.

File: include/mbedtls/ecdh.h [hunk 5]
```diff
@@ -40,6 +40,28 @@
...
+ * no difference, since all library functions can work with either format,
+ * except that the new format is incompatible with MBEDTLS_ECP_RESTARTABLE.
+
+ * The new format used when this option is disabled is smaller
+ * (56 bytes on a 32-bit platform). In future versions of the library, it
+ * will support alternative implementations of ECDH operations.
+ * The new format is incompatible with applications that access
+ * context fields directly and with restartable ECP operations.
+ *
+ * This option has no effect if #MBEDTLS_ECDH_C is not enabled.
```
Comment (ronald-cron-arm) [inline include/mbedtls/ecdh.h:55]: This last sentence had sense in config.h but not really here thus I would remove it.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. I just have two improvement suggestions.

**Comment by tom-daubney-arm** [comment]
Thank you, changes have now been made.

**Review by ronald-cron-arm — APPROVED** [review]

**Review by mpg — APPROVED** [review]
Looks good to me as well, Thanks Ronald for noticing the documentation of `ECP_RESTARTABLE` needed updating.

**Pull request merged**